pulsar-broker/src/test/java/org/apache/pulsar/broker/service/GracefulExecutorServicesShutdownTest.java

```diff
@@ -120,25 +121,28 @@ public void shouldWaitForExecutorToTerminate() throws ExecutionException, Interr
 
 
     @Test
-    public void shouldTerminateWhenFutureIsCancelled() throws InterruptedException {
+    public void shouldTerminateWhenFutureIsCancelled() throws InterruptedException, ExecutionException {
         // given
         GracefulExecutorServicesShutdown shutdown = GracefulExecutorServicesShutdown.initiate();
         shutdown.timeout(Duration.ofMillis(15000));
         ExecutorService executorService = mock(ExecutorService.class);
         when(executorService.isShutdown()).thenReturn(true);
         AtomicBoolean terminated = new AtomicBoolean();
-        AtomicBoolean awaitTerminationInterrupted = new AtomicBoolean();
+        CompletableFuture<Boolean> awaitTerminationInterrupted = new CompletableFuture<>();
         when(executorService.isTerminated()).thenAnswer(invocation -> terminated.get());
+        CountDownLatch awaitingTerminationEntered = new CountDownLatch(1);
         when(executorService.awaitTermination(anyLong(), any())).thenAnswer(invocation  -> {
             long timeout = invocation.getArgument(0);
             TimeUnit unit = invocation.getArgument(1);
+            awaitingTerminationEntered.countDown();
```

Comment:
move `awaitingTerminationEntered.countDown();` into try block

Reply:
nice, I happen to be about to solve this problem

Reply:
> move `awaitingTerminationEntered.countDown();` into try block

Does that make a difference?

Reply:
1. awaitingTerminationEntered.countDown() and if the thread do not run into try block.
2. future.cancel(false);
then awaitTerminationInterrupted maybe never complete.

Reply:
What caused the flakiness was that this condition caused the execution to never get to the Thread.sleep when the thread was already interrupted:
https://github.com/apache/pulsar/blob/a1cebdb3e7cc3b8ee3ce567058182e7d31c762d2/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/GracefulExecutorServicesTerminationHandler.java#L112

This is fixed with the use of the CountDownLatch.

Reply:
I don't see a reason to move `awaitingTerminationEntered.countDown();` into the try block since that doesn't make a difference.